aopy/data/bmi3d.py

```diff
@@ -889,6 +890,84 @@ def tabulate_behavior_data_center_out(preproc_dir, subjects, ids, dates, metadat
     df = pd.concat([df, new_df], ignore_index=True)
     return df
 
+def tabulate_behavior_data_tracking_task(preproc_dir, subjects, ids, dates, metadata=[], 
+                                      df=None, include_before_initiation=False, include_ramps=False):
+    '''
+    Wrapper around tabulate_behavior_data() specifically for tracking task experiments. 
+    Makes use of the task codes saved in `/config/task_codes.yaml` to automatically 
+    assign event codes for trial start, trial end, reward, penalty. 
+    Trial start can optionally include the time before trial initiation (i.e. hold).
+    
+    Args:
+        preproc_dir (str): base directory where the files live
+        subjects (list of str): Subject name for each recording
+        ids (list of int): Block number of Task entry object for each recording
+        dates (list of str): Date for each recording
+        metadata (list, optional): list of metadata keys that should be included in the df
+        df (DataFrame, optional): pandas DataFrame object to append. Defaults to None.
+        include_before_initiation (bool, optional): If True, trials begin as soon as center target 
+            and trajectory appear. Otherwise trials begin once hold is initiated. Default False.
+        include_ramps (bool, optional): If True, trajectory_times correspond the start and end of 
+            the trajectory including the ramp up and down. Otherwise trajectory_times excludes 
+            ramp periods. Default False.
+
+    Returns:
+        pd.DataFrame: pandas DataFrame containing the concatenated trial data
+    '''
+    # Use default "trial" definition
+    config_dir = files('aopy').joinpath('config')
+    params_file = as_file(config_dir.joinpath('task_codes.yaml'))
+    with params_file as f:
+        task_codes = yaml_read(f)[0]
+    trial_end_codes = [task_codes['REWARD'], task_codes['TIMEOUT_PENALTY'], task_codes['HOLD_PENALTY'], task_codes['OTHER_PENALTY']]
+    reward_codes = [task_codes['REWARD']]
+    
+    if include_before_initiation:
+        trial_start_codes = [task_codes['CENTER_TARGET_ON']]
+        penalty_codes = [task_codes['TIMEOUT_PENALTY'], task_codes['HOLD_PENALTY'], task_codes['OTHER_PENALTY']]
+    else:
+        trial_start_codes = [task_codes['TRIAL_START']]
+        penalty_codes = [task_codes['HOLD_PENALTY'], task_codes['OTHER_PENALTY']]
```

Comment:
This is fine and similar to the way i set up the center_out function. But i've been meaning to simplify it so that it always includes the entire trial (from when the center target turns on until the trial end code) so that the dataframes are more consistent and you never have to remake them. I suggest you do the same at some point. It just requires that you be more thoughtful about selecting trial start and end times later, and we should write some functions that take in one trial worth of event codes and spit out the indices that define various forms of sub-trial, like a center-out reach or a tracking-in period, etc.